src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -57,6 +56,106 @@
 #include "utilities/events.hpp"
 #include "utilities/preserveException.hpp"
 
+void MonitorList::add(ObjectMonitor* m) {
+  for (;;) {
+    ObjectMonitor* head = Atomic::load(&_head);
+    m->set_next_om(head);
+    if (Atomic::cmpxchg(&_head, head, m) == head) {
+      break;
+    }
+  }
+  size_t count = Atomic::add(&_count, 1u);
+  if (count > max()) {
+    Atomic::inc(&_max);
+  }
+}
+
+size_t MonitorList::count() const {
+  return Atomic::load(&_count);
+}
+
+size_t MonitorList::max() const {
+  return Atomic::load(&_max);
+}
+
+// Walk the in-use list and unlink (at most MonitorDeflationMax) deflated
+// ObjectMonitors. Returns the number of unlinked ObjectMonitors.
+size_t MonitorList::unlink_deflated(Thread* self, LogStream* ls,
+                                    elapsedTimer* timer_p,
+                                    GrowableArray<ObjectMonitor*>* unlinked_list) {
+  size_t unlinked_count = 0;
+  ObjectMonitor* prev = NULL;
+  ObjectMonitor* head = Atomic::load_acquire(&_head);
+  ObjectMonitor* m = head;
+  while (m != NULL) {
+    if (m->is_being_async_deflated()) {
+      // Find next live ObjectMonitor.
+      ObjectMonitor* next = m;
+      while (next != NULL && next->is_being_async_deflated()) {
+        ObjectMonitor* next_next = next->next_om();
+        unlinked_count++;
+        unlinked_list->append(next);
+        next = next_next;
+        if (unlinked_count >= (size_t)MonitorDeflationMax) {
+          // Reached the max so bail out on the gathering loop.
+          break;
+        }
+      }
+      if (prev == NULL) {
+        ObjectMonitor* prev_head = Atomic::cmpxchg(&_head, head, next);
+        if (prev_head != head) {
+          // Find new prev ObjectMonitor that just got inserted.
+          for (ObjectMonitor* n = prev_head; n != m; n = n->next_om()) {
+            prev = n;
+          }
+          prev->set_next_om(next);
+        }
+      } else {
+        prev->set_next_om(next);
+      }
+      if (unlinked_count >= (size_t)MonitorDeflationMax) {
+        // Reached the max so bail out on the searching loop.
+        break;
+      }
+      m = next;
+    } else {
+      prev = m;
+      m = m->next_om();
+    }
+
+    if (self->is_Java_thread() &&
+        SafepointMechanism::should_process(self->as_Java_thread())) {
+      // A safepoint/handshake has started.
+      if (ls != NULL) {
+        timer_p->stop();
+        ls->print_cr("pausing unlinking: unlinked_count=" SIZE_FORMAT ", in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     unlinked_count, ObjectSynchronizer::in_use_list_ceiling(), count(), max());
+      }
+
+      // Honor block request.
+      ThreadBlockInVM tbivm(self->as_Java_thread());
```

Comment:
ThreadBlockInVM is generally used to wrap blocking code, not to cause the current thread to block (which it will do as a side-effect if a safepoint/handshake is requested). Surely this should just be call to `process_if_requested` (or the new `process_if_requested_with_exit_check`)?

Reply:
This kind of use of ThreadBlockInVM predates this changeset so while
the location is new, then code style is old, very old... I'll hold off changing
this for now.

Reply:
I'd rather see ThreadBlockInVM as the convention of allowing the thread to block if a safepoint is requested.  The calls like process_if_requested are becoming alphabet soup and keep changing, so having TBIVM is better in my opinion.
That said, this is a strange usage.  This code appears three times.  It should be a function like allow_safepoint_block(LogStream* ls, timer), with some comment above.  Then it's clear that it's checking for a safepoint in a loop that could take a long time and the logging is ancillary.

Reply:
I previously looked at refactoring the three locations where
`ThreadBlockInVM` is used. The problem with the refactoring
is that the log messages and the parameters have some
differences and some commonalities. Each of these logging
sites is trying to communicate some local context that is
unique to that call site along with some global context that
might have changed from call site to call site.

I'll take another look at refactoring shortly and will let you
know what I come up with.

Reply:
I refactored common ThreadBlockInVM code block into ObjectSynchronizer::chk_for_block_req().
Please let me know if you're okay with it.